src/physics/boundaryConditions/PeriodicBoundary.cpp

```diff
@@ -0,0 +1,230 @@
+        // -> That way we make sure to only transform cells on edges or corners once
+        if (boundarySides[0] != position)
+            continue;
+
+        // Only look at cells that are affected only by periodic bounds -> otherwise the other bound
+        // will hold
+        for (Position boundSide : boundarySides) {
+            // Bounds must be periodic
+            if (boundaryConfig.boundaryMap.at(boundSide) != BoundaryType::PERIODIC)
+                continue;
+        }
+
+        // Check if the translations have been generated already
+        if (translationMap.find(boundarySides) != translationMap.end())
+            continue;
+
+        // Add the necessary translations
+        fillTranslationMap(
+            boundarySides, { boundarySides[0] }, cellGrid); // Always do the single inner one
+        if (boundarySides.size() >= 2) {
+            fillTranslationMap(boundarySides, { boundarySides[1] }, cellGrid);
+            fillTranslationMap(boundarySides, { boundarySides[0], boundarySides[1] }, cellGrid);
+        }
+        if (boundarySides.size() == 3) {
+            fillTranslationMap(boundarySides, { boundarySides[2] }, cellGrid);
+            fillTranslationMap(boundarySides, { boundarySides[0], boundarySides[2] }, cellGrid);
+            fillTranslationMap(boundarySides, { boundarySides[1], boundarySides[2] }, cellGrid);
+            fillTranslationMap(boundarySides, boundarySides, cellGrid);
+        }
+        if (boundarySides.size() > 3) {
+            spdlog::error(
+                "A cell has been found to have more than 3 associated sides. This is a bug");
+        }
+    }
+}
+
+void PeriodicBoundary::preUpdateBoundaryHandling(Simulation& simulation)
+{
+    const LennardJonesDomainSimulation& LGDSim =
+        static_cast<const LennardJonesDomainSimulation&>(simulation);
+
+    // reset the insertion index, to reuse already created particles
+    insertionIndex = 0;
+    for (CellIndex boundaryCellIndex : LGDSim.getGrid().boundaryCellIterator(position)) {
+        std::vector<Position> boundarySides = coordinateToPosition(
+            boundaryCellIndex, LGDSim.getGrid().getGridDimensions(), !is2D, false);
+
+        // Only if translations need to be applied i.e. there exists an entry
+        if (translationMap.find(boundarySides) == translationMap.end())
+            continue;
+
+        for (auto& particle :
+             LGDSim.getGrid()
+                 .cells[boundaryCellIndex[0]][boundaryCellIndex[1]][boundaryCellIndex[2]]
+                 ->getParticles()) {
+            for (PeriodicBoundShifts& shifts : translationMap.at(boundarySides)) {
+                std::array<double, 3> pos = particle.get().getX();
+                std::array<double, 3> haloPosition = pos + shifts.first;
+
+                // Ugly but we cannot simply add as size_t != int
+                CellIndex haloCellIndex { (size_t)((int)boundaryCellIndex[0] + shifts.second[0]),
```

Comment:
```suggestion
                CellIndex haloCellIndex { static_cast<size_t>(static_cast<int>(boundaryCellIndex[0]) + shifts.second[0]),
```
In general, we always prefer C++ style type casting, even though this sadly makes this expression even uglier. See [here](https://stackoverflow.com/questions/1609163/what-is-the-difference-between-static-cast-and-c-style-casting).